File: Tree/100-538-convertBST.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
    
    def preOrder(self,root):
        if not root:
            return
        print(root.val)
        self.preOrder(root.left)
        self.preOrder(root.right)
class Solution:
    # num=0
    
    def convertBST(self, root):
        if not root:
            return None
        num=0
        def dfs(root):
            if not root:
                return None
            nonlocal num
            dfs(root.right)
            num+=root.val
            root.val=num
            dfs(root.left)
            
            return root
        return dfs(root)
    

one=TreeNode(1)
two=TreeNode(2)
three=TreeNode(3)
four=TreeNode(4)
five=TreeNode(5)

three.left=two
two.left=one

three.right=four
four.right=five

three.preOrder(three)
newRoot=Solution().convertBST(three)
print("----------------------")
newRoot.preOrder(newRoot)

File: Tree/100-538-convertBST.cpp
#include "treeNode.h"
#include<iostream>

using std::cout;

TreeNode* dfs(TreeNode *root,int &num)
{
    if(!root)
        return nullptr;
    dfs(root->right,num);
    num+=root->m_val;
    root->m_val=num;
    dfs(root->left,num);
    return root;
}

TreeNode *convertBST(TreeNode *root)
{
    if(!root)
        return nullptr;
    
    int num=0;
    dfs(root,num);
    return root;
}

int main()
{
    TreeNode *one=new TreeNode(1);
    TreeNode *two=new TreeNode(2);
    TreeNode *three=new TreeNode(3);
    TreeNode *four=new TreeNode(4);
    TreeNode *five=new TreeNode(5);

    three->left=two;
    two->left=one;

    three->right=four;
    four->right=five;
    three->printTreeNodePreOrder(three);
    cout<<"----------------\n";
    TreeNode *newRoot=convertBST(three);
    newRoot->printTreeNodePreOrder(newRoot);
}